gallery_dl/extractor/generic.py

```diff
@@ -0,0 +1,207 @@
+        """
+        imageurl_pattern_src = r"""(?ix)
+            <(?:img|video|source)\s.*?      # <img>, <video> or <source>
+            src(?:set)?=["']?               # src or srcset attributes
+            (?P<URL>[^"'\s>]+)              # url
+            """
+
+        """
+        2: Look anywhere for urls containing common image/video extensions
+
+        The list of allowed extensions is borrowed from the directlink.py
+        extractor; other could be added, see
+        https://en.wikipedia.org/wiki/List_of_file_formats
+
+        We don't allow relative urls here, i.e. generic strings containing an
+        image filename (such as in <... alt="image.jpg">) because they often
+        get resolved into non-existing absolute urls, leading to 404 errors
+        that might confuse the user.
+
+        Compared to the "pattern" class variable, here we must exclude also
+        other special characters (space, ", ', >), since we are looking for
+        urls in html tags.
+
+        Note that we use the x flag, so spaces and "#" must be
+        backslash-quoted, unless they are in a character class ([]).
+        """
+        imageurl_pattern_ext = r"""(?ix)
+            (?:https?://[^/?&#"'>\s]+)          # scheme + domain
+            (?:[^?&#"'>\s]+)                    # path until dot+ext
+            \.(?:jpe?g|jpe|png|gif
+                 |web[mp]|mp4|mkv|og[gmv]|opus) # dot + image/video extensions
+            (?:\?[^"'>\s]*)?                    # optional query and fragment
+            """
+
+        imageurls_src = re.findall(imageurl_pattern_src, page)
+        imageurls_ext = re.findall(imageurl_pattern_ext, page)
+        imageurls = imageurls_src + imageurls_ext
+
+        # Resolve relative urls
+        #
+        # Some of the urls may be relative, so we resolve them prefixing them
+        # either by the page root url (self.root) if the relative url starts
+        # with "/", or by a proper "base" url if the relative url doesn't start
+        # with "/"
+
+        # If the page contains a <base> element, use it as base url
+        basematch = re.search(
+            r"(?i)(?:<base\s.*?href=[\"']?)(?P<url>[^\"' >]+)", page)
+        if basematch:
+            self.baseurl = basematch.group('url')
+        else:
+            # Use self.url if it doesn't end with an extension
+            if os.path.splitext(self.url)[1] == "":
+                self.baseurl = self.url
+            # Otherwise, strip the last part
+            else:
+                self.baseurl = os.path.dirname(self.url)
+
+        # Really resolve relative urls
+        absimageurls = []
+        for u in imageurls:
+            if u.startswith('http'):
+                absimageurls.append(u)
+            elif u.startswith('/'):
+                absimageurls.append(self.root + '/' + u)
+            else:
+                absimageurls.append(self.baseurl + '/' + u)
+
+        # Remove duplicates
+        absimageurls = set(absimageurls)
```

Comment:
This reorders all image URLs on Python 3.4 and 3.5.
I don't think that's too important, but there are ways to remove duplicates without putting everything in a set().

Reply:
I'm not sure image urls order is really important, usually you end up with a directory of image files, and what matters is the file name; of course the image url order would matter if you used `num` to build a custom `filename_fmt`.

Anyway, even if we removed duplicates by preserving image url order in the `imageurls` list, at the moment this wouldn't reflect the order of image urls in the page, since `imageurls` is the product of two passes on the page, one for each search strategy (`imageurl_pattern_src` and `imageurl_pattern_ext`); but of course we could merge them into one big alternative regex and only make one pass (I originally divided it in two mainly for ease of testing).

So we have two options:
1. Status quo: don't care about image urls order
2. Refactor: merge imageurl search strategies into one regex, and remove duplicates preserving imageurl order

I would go for 1. for now, and leave 2. as an enhancement for the future

Reply:
Uhm, wait: do I really need to add extractor code to remove duplicate image urls? Can't I just set `extractor.generic.image-unique` in the extractor, and gallery-dl will take care of removing duplicates? Or am I missing something?